### PR TITLE
Add missing Request Body for DELETE APIs re: site issue #905

### PR DESCRIPTION
**Pull request by jaywood128**
…

re: https://github.com/FusionAuth/fusionauth-site/issues/905

### CODE REVIEW

**Review by mooreds — CHANGES_REQUESTED** [review]
A few changes. Biggest q is should we be documenting this or prohibiting the behavior.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:9]: Seeing it for `/api/two-factor/start` and `GET /api/two-factor/status/{twoFactorTrustId}`. Only see `code` and `methodId` for existing request parameters but perhaps I am looking in the wrong place? https://fusionauth.io/docs/v1/tech/apis/two-factor#request-2

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/two-factor.adoc:151]: Nice catch! I will have to test to confirm I can also provide a JWT. Will add once the test is confirmed.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/two-factor.adoc:155]: I audited all of the APIs and this was the only one that I found that was undocumented. @mooreds my understanding of this ticket was to ensure all APIs that take a Request Body are documented. I can't imagine it would provide much in the way of convenience for endpoint users to add this to a bunch of DELETE endpoints, and I couldn't find any evidence to show that URL parameters are more secure than a request body. I would say document the ones we are already exposing. I don't think there would be any concerns about size limits in a request body that takes these two parameters unless it was some kind of bulk operation, in which case it probably already takes a Request Body. Perhaps I am missing something in terms of security and maybe keeping the APIs consistent?

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:12]: Just ran through this again with the following request body and MFA was disabled. 

`{
  "code": "9N8MR-9B2PL",
  "methodId": "4GFV"
}`

I believe the `methodId` is used on the validateDelete(), and it's also returned on the `result` which also gets used to pass it into `io.fusionauth.app.action.api.user.TwoFactorAction#userService.disableTwoFactor ` line 73.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/two-factor.adoc:151]: Just re-tested with a JWT and a Request Body. Returned a 200 and I saw the MFA get removed in the UI for the user.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (mooreds) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:9]: I'd look in the validation method. I see that an optional `applicationId` is accepted.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (mooreds) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:12]: okay

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (mooreds) [reply site/docs/v1/tech/apis/two-factor.adoc:155]: okay, if this is a one-off, fine documenting it for backwards compatibility. Might be worth raising the bigger q with the engineering team about any other delete endpoints going forward.

**Comment by mooreds** [comment]
Still not sure about the applicationId, I think that needs to be added.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:13]: Correct, the `appliationId` is optionally accepted, my mistake.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [inline site/docs/v1/tech/apis/two-factor.adoc:167]: When using a JWT, you don't provide a `userId` on the URL - the JWT provides the `userId` by using the `sub` claim.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:6]: The application id for email templates and events. - found this description for `io.fusionauth.api.service.user.UserService#validateDisableTwoFactor`. I don't see `result.application` being used elsewhere in the method so I am confused as to how it's used. Perhaps this value is used in state to log a user in after they've completed MFA? I also see it being used the the validateEnableMFA method, so I will document it there as well once I get the description is accurate.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/two-factor.adoc:167]: 👍

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:6]: `applicationId` is mentioned here as well but still now sure how it's being used for disabling MFA. Why would an email be sent or used in an email templated to disable MFA? https://fusionauth.io/docs/v1/tech/apis/emails#send-an-email

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:6]: It also could be referring to the `applicationId` that was provided to start the MFA process and the one that is then provided in the email templates. https://fusionauth.io/docs/v1/tech/apis/two-factor#request-4

`[field]#applicationId# [type]#[UUID]# [optional]#Optional#::
The application id that was provided to Start Multi-Factor Authentication used for email templates and events.`

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:13]: @mooreds I want to confirm I have addressed concerns about documenting the `applicationId` before I merge. Thanks!

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (mooreds) [inline site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:6]: This doesn't quite explain what this is useful for.

Can you explain a bit more? Like, why would I pass the `applicationId` parameter?

Also, we captialize "Id"

**Review by mooreds — CHANGES_REQUESTED** [review]

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:6]: @mooreds the `applicationId` is used for logging events and allow the usage of application-specific email templates for MFA, if configured at the application level. I tested this by including the `applicationId` at the top level of the request, and configured my custom application-specific email MFA template but the tenant application was used so this might be a bug.

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: @mooreds - as discussed a section on how to override the Tenant email template, in favor of an application specific one will have to be documented, as well as one for forgot password, as mentioned here: https://github.com/fusionauth/fusionauth-issues/issues/1005

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: @mooreds I think it would make sense to add the override section here, unless you think another place makes more sense? https://fusionauth.io/docs/v1/tech/email-templates/email-templates

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (mooreds) [inline site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:5]:
```suggestion
The Id of the application that has been configured with a Two Factor Method Added email template. Please see the section on using link:#XXX[application specific email templates to learn more]. This Id is also used for events.
```

Where's the link? Doesn't make sense to add this until the section is written.

File: site/docs/v1/tech/apis/_two-factor-enable-request-body.adoc
Comment (mooreds) [inline site/docs/v1/tech/apis/_two-factor-enable-request-body.adoc:5]: Same comments as above.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (mooreds) [reply site/docs/v1/tech/apis/two-factor.adoc:155]: I think we can resolve this convo.

**Review by mooreds — CHANGES_REQUESTED** [review]

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (mooreds) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: It is a premium feature, correct (only available with a license)? If is I'd add a separate page, as we'll want it to show up in the premium features section. Fine to link to it from https://fusionauth.io/docs/v1/tech/email-templates/email-templates

If not, https://fusionauth.io/docs/v1/tech/email-templates/email-templates is fine. Still think a separate page would make sense.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:5]: A user likely wouldn't know how to use the `applicationId` parameter without a guide/video showing them how to override the Tenant Email Template. I'll make a guide, link that pr to this one, and then provide the link here to the guide.  @mooreds how does that sound?

Comment (mooreds) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:5]: That sounds good. So the plan is to not merge this until the guide is done in a sep PR?

Works for me.

Comment (mooreds) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:5]: @jaywood128

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: Opened the pull request for the guide here: https://github.com/FusionAuth/fusionauth-site/pull/1882 and added a link and more consistent descript for `applicationId` below. @mooreds

**Comment by mooreds** [comment]
Plan is to wait until #1882 is merged. So no need to review this until then, correct @jaywood128 ?

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (robotdan) [inline site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:14]: can we use `two-factor method` here since we are in the Two Factor API?

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (robotdan) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:16]: We should also add that when a recovery code is used to disable two-factor, all two-factor methods will be removed. So, for example, even if you have 10 various two-factor methods, if you disable one using a recovery code, all will be removed.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [reply site/docs/v1/tech/apis/two-factor.adoc:155]: We do have other APIs that take a request body or query parameters. The two reasons for this are: (you nailed them)

- Overcome request size limits. For example, bulk user delete - appending thousands of `userIds=` to the query string will blow out the request URL. 
- Can be more confidential 

In general, if an API supports using the request body or query parameters, the client lib version will use the body. This is because it is simpler to write, and requires less code maintenance. 

I would generally speaking say we prefer the request body, but I don't think we should deprecate anything since a using a request body on a `DELETE` request may not always be possible for a client. 

Using the request body is a nice to have if you can do it. The w3 spec suggests that `GET` and `DELETE` methods don't take a request body, but in practice this is done frequently in REST because of the limitations mentioned above.

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [inline site/docs/v1/tech/apis/two-factor.adoc:163]: Let's remove `with a UserId` here, and don't capitalize `With` 

> Disable Multi-Factor Authentication using a JWT

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [inline site/docs/v1/tech/apis/two-factor.adoc:151]: > Disable Multi-Factor Authentication by User Id

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [inline site/docs/v1/tech/apis/two-factor.adoc:175]: > Disable Multi-Factor Authentication using a JWT

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:9]: The name of the endpoint is `Disable Multi-Factor Authentication With a UserId`. Although it's the Two Factor API, most of the language on the page uses Multi-Factor > Two Factor. I copy and pasted this from the request parameters so I will be sure to clean make it apply to TOTP, Email, and SMS.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:14]: Okay, so replace all mentions of Mult-Factor with `two-factor`? This would follow the path for the API as well. Once you confirm, I will go ahead and make those changes, if that is what you would prefer. I agree that it would be a better convention to follow.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:9]: I can change this to favor `two-factor` over `multi-factor but I wanted to confirm the change before I did.

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: That description for `code` appears 4 times on the Two Factor Page so I would have to update it. Here's my new description:


_A valid Two-Factor verification code. The User will obtain this code by either a Time-based one-time password (TOTP), Email, or SMS to prove they have access to codes._

The `applicationId` can be included in the Request Body - this is what I originally tested and found doing an audit. The request parameters are URL parameters. Our code is pulling the applicationId off of the request body. I will re-test to confirm again to make sure I am getting it right because it has been a few weeks.

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: After reviewing the Multi-Factor/Two-Factor API, IMO, I think it makes more sense to keep Multi-Factor. When you enable multi-factor, you can select 1 of 3 values: authenticator, email, SMS. Since multiple methods can be enabled, I think it makes more sense to keep the 73 mentions of Multi-Factor over two-factor.

Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: @robotdan - I'm sticking with `Multi-Factor` over `multi-factor` - otherwise, I would have to update all other fields. Would you rather me update all `Mutli-Factor` -> `multi-factor` or continue to use `Multi-Factor`?

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:14]: The API is the Multi-Factor/TwoFactor api but the path is `two-factor` so it's definitely confusing on which to choose. `Multi-Factor ` was used by all fields so I stuck with that one.

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (jaywood128) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc:16]: @robotdan this would be the default behavior? Sounds like a modification/enhancement of the API, should I make a new issue to track this?

**Comment by robotdan** [comment]
I think you're still missing the `applicationId` on the `DELETE` request when using query parameters?

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (robotdan) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: made some edits to use `multi-factor`

File: site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc
Comment (robotdan) [reply site/docs/v1/tech/apis/_two-factor-disable-request-body.adoc]: made some edits to use `multi-factor`

File: site/docs/v1/tech/apis/two-factor.adoc
Comment (robotdan) [reply site/docs/v1/tech/apis/two-factor.adoc]: I made an edit here I think.

**Comment by jaywood128** [comment]
> I think you're still missing the `applicationId` on the `DELETE` request when using query parameters?

I tested this with Postman and got a 500. As soon as I removed `applicationId` from the qs, the request worked. I will write a test to confirm.

**Review by robotdan — COMMENTED** [review]
We should also be taking an `applicationId` the APIs that are using query parameters. If the code isn't using it, that would be a bug, so please open a GH issue to document this bug so that we can fix it. 

And fix the doc to match how we want it to work.these APIs.

**Comment by jaywood128** [comment]
> We should also be taking an `applicationId` the APIs that are using query parameters. If the code isn't using it, that would be a bug, so please open a GH issue to document this bug so that we can fix it.
> 
> And fix the doc to match how we want it to work.these APIs.

The Disable Multi-Factor API endpoint should take an `applicationId` - I will link a bug for that for fusionauth-issues, if so. Also, if this is correct, I will also make a fusionauth-site issue to document the `applicationId` for the Disable Multi-Factor Endpoint. 

Does this mean any endpoint that takes a query string parameter should take an `applicationid`. If so, that would better inform the audit, and I can take a look through to make sure each endpoint is doing that.

**Comment by jaywood128** [comment]
> We should also be taking an `applicationId` the APIs that are using query parameters. If the code isn't using it, that would be a bug, so please open a GH issue to document this bug so that we can fix it.
> 
> And fix the doc to match how we want it to work.these APIs.

Doc issue with `fusionauth-issue` linked : https://github.com/FusionAuth/fusionauth-issues/issues/2153

**Pull request merged**
